File: Sources/PulseUI/ViewModels/ModuleModel.swift
import Foundation
import Observation
import PulseCore

/// What a module card can be showing at any moment.
public enum ModulePhase<Payload: Codable & Equatable & Sendable>: Equatable {
    case loading
    case loaded(ProviderResult<Payload>)
    case failed(String)
}

/// Observation-based view model for one dashboard module.
///
/// Generic over the provider, so every module gets identical semantics:
/// load → fresh, or cached-marked-stale, or a readable failure. The view
/// just switches over `phase`.
@Observable
@MainActor
public final class ModuleModel<Provider: DataProvider> {
    public private(set) var phase: ModulePhase<Provider.Payload> = .loading

    private let provider: Provider
    private let cache: PayloadCache<Provider.Payload>

    public init(provider: Provider, cache: PayloadCache<Provider.Payload>) {
        self.provider = provider
        self.cache = cache
    }

    public func load(now: Date = Date()) async {
        do {
            let result = try await provider.fetchCachingLastGood(cache: cache, now: now)
            phase = .loaded(result)
        } catch {
            phase = .failed("Couldn't load \(provider.title.lowercased()). Pull to retry.")
        }
    }
}

File: android/core/src/main/kotlin/com/pulse/core/ModuleModel.kt
package com.pulse.core

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** What a module card can be showing at any moment — the Kotlin mirror of the
 *  Swift `ModulePhase` enum, as a sealed hierarchy. */
sealed interface ModulePhase<out T> {
    data object Loading : ModulePhase<Nothing>
    data class Loaded<T>(val result: ProviderResult<T>) : ModulePhase<T>
    data class Failed(val message: String) : ModulePhase<Nothing>
}

/**
 * View-model for one dashboard module. Generic over the provider, so every
 * module gets identical semantics: load → fresh, or cached-marked-stale, or a
 * readable failure. The UI just collects [phase].
 *
 * `StateFlow` is Kotlin's Observation: the view observes it, the model owns it.
 */
class ModuleModel<T>(
    private val provider: DataProvider<T>,
    private val cache: PayloadCache<T>,
) {
    private val _phase = MutableStateFlow<ModulePhase<T>>(ModulePhase.Loading)
    val phase: StateFlow<ModulePhase<T>> = _phase.asStateFlow()

    suspend fun load(now: Long) {
        _phase.value = try {
            ModulePhase.Loaded(provider.fetchCachingLastGood(cache, now))
        } catch (_: Exception) {
            ModulePhase.Failed("Couldn't load ${provider.title.lowercase()}. Pull to retry.")
        }
    }
}
